src/blr/cli/extractbarcode.py

```diff
@@ -3,45 +3,110 @@
+        logger.info(f'Writing interleaved fastq to stdout.')
+    elif output1 and not output2:
+        interleaved = True
+        logger.info(f'Writing interleaved fastq to file {output1}.')
+
+    return output1, output2, interleaved
+
+
 def main(args):
-    logger.info('Starting')
+    logger.info(f'Starting')
+
     progress = BLR.ProgressReporter('Read pairs processed', 1000000)
-    generator = BLR.FileReader(args.r1, args.r2)
-    with open(args.out_r1, 'w') as openr1, open(args.out_r2, 'w') as openr2:
-        for read1, read2 in generator.fastqPairedReader():
 
-            # Adjusting for BC
-            bc_seq = read1.seq[:20]
-            read1.seq = read1.seq[20:]
-            read1.qual = read1.qual[20:]
+    inputfile1, inputfile2, input_interleaved = input_files(args.input, args.interleaved)
+
+    reader = dnaio.open(inputfile1, file2=inputfile2, interleaved=input_interleaved, mode="r", fileformat="fastq")
+
+    outputfile1, outputfile2, output_interleaved = output_files(args.output1, args.output2)
+
+    writer = dnaio.open(outputfile1, file2=outputfile2, interleaved=output_interleaved, mode="w", fileformat="fastq")
+
+    for read1, read2 in reader:
+        # Adjusting for BC
+        bc_seq = read1.sequence[:20]
+        read1.sequence = read1.sequence[20:]
+        read1.qualities = read1.qualities[20:]
 
-            # Header parsing
-            name_and_pos_r1, read_and_index_r1 = read1.header.split(maxsplit=1)
-            name_and_pos_r2, read_and_index_r2 = read2.header.split(maxsplit=1)
+        # Header parsing
+        name_and_pos_r1, read_and_index_r1 = read1.name.split(maxsplit=1)
+        name_and_pos_r2, read_and_index_r2 = read2.name.split(maxsplit=1)
 
-            # Save header to read instances
-            read1.header = name_and_pos_r1 + '_' + bc_seq + ' ' + read_and_index_r1
-            read2.header = name_and_pos_r2 + '_' + bc_seq + ' ' + read_and_index_r2
+        # Save header to read instances
+        read1.name = name_and_pos_r1 + '_' + bc_seq + ' ' + read_and_index_r1
+        read2.name = name_and_pos_r2 + '_' + bc_seq + ' ' + read_and_index_r2
 
-            # Write to out
-            openr1.write(read1.fastq_string())
-            openr2.write(read2.fastq_string())
+        # Write to out
+        writer.write(read1, read2)
 
-            # Progress reporting
-            progress.update()
+        # Progress reporting
+        progress.update()
 
-    generator.close()
-    logger.info('Finished')
+    reader.close()
+    writer.close()
+    logger.info(f'Finished')
 
 
 def add_arguments(parser):
-    parser.add_argument("r1", help="Read 1 fastq file")
-    parser.add_argument("r2", help="Read 2 fastq file")
-    parser.add_argument("out_r1", help="Read 1 output")
-    parser.add_argument("out_r2", help="Read 2 output")
+    # Files
+    parser.add_argument("input", nargs='*',
+                        help="Input paired fastq for read1 and read2. Use '-' if using stdin.")
```

Comment:
Since the function at least requires one argument I would write it as 
```
parser.add_argument("input1", help="Input paired fastq for read1 or interleaved file. Use '-' if using stdin.")
parser.add_argument("input2", nargs='?', help="Input paired fastq for read2. Use '-' if using stdin.")
```
This will replace all error handling in the ```input_files``` function and give standardized error messages from argparse for faulty user input.

Reply:
Subsequently one could remove the ```input_file``` function with 
```
if not args.input2: 
    interleaved = input1
    input1 = None
```
in the main function instead. One could also write ```if not args.input2 and args.interleaved:``` for error handling if we believe users might try and input single-end sequences.